doc/CodeSize.md

```diff
@@ -0,0 +1,260 @@
+
+## Code Style: tips for keeping Rust code small
+
+#### When to use generic types with trait bounds versus trait objects (`dyn`)
+
+Polymorphic structs and functions are one of the biggest sources of bloat in Rust
+binaries -- use of generic types can lead to bloat from monomorphization, while
+use of trait objects introduces vtables into the binary and limits opportunities
+for inlining.
+
+Use `dyn` when the function in question will be called with multiple concrete types;
+otherwise code size is increased for every concrete type used (monomorphization).
+
+```rust
+fn set_gpio_client(&dyn GpioClientTrait) -> Self {//...}
+
+// elsewhere
+let radio: Radio = Radio::new();
+set_gpio_client(&radio);
+
+let button: Button = Button::new();
+set_gpio_client(&button);
+```
+
+Use generics with trait bounds when the function is only ever called with a single public type;
+This reduces code size and run time cost. This increases source code complexity for decreased
+image size and decreased clock cycles used.
+
+```rust
+// On a given chip, there is only a single FlashController. We use generics so
+// that there can be a shared interface by all FlashController's on different chips,
+// but in a given binary this function will never be called with multiple types.
+impl<'a, F: FlashController> StorageDriverBackend<'a, F> {
+    pub fn new(
+        storage: &'a StorageController<'a, F>,
+    ) -> Self { ... }
+
+```
+
+Similarly, only use const generics when there will not be monomorphization, or if the body of the
+method which would be monomorphized is sufficiently small that it will be inlined anyways.
+
+
+#### Panics
+Panics add substantial code size to Tock binaries -- on the order of 50-75 bytes per panic. Returning
+errors is much cheaper than panicing, and also produces more dependable code. Whenever possible, return
+errors instead of panicing. Often, this will not only mean avoiding explicit panics: many core library
+functions panic internally depending on the input.
+
+The most common panics in Tock binaries are from array accesses, but these can often be ergonomically
+replaced with result-based error handling:
+
+```rust
+// BAD: produces bloat
+fn do_stuff(&mut self) -> Result<(), ErrorCode> {
+    if self.my_array[4] == 7 {
+        self.other_array[3] = false;
+        Ok(())
+    } else {
+        Err(ErrorCode::SIZE)
+    }
+}
+```
+```rust
+// GOOD
+fn do_stuff(&mut self) -> Result<(), ErrorCode> {
+    if self.my_array.get(4).ok_or(ErrorCode::FAIL)? == 7 {
+        *(self.other_array.get_mut(3).ok_or(ErrorCode::FAIL)?) = false;
+        Ok(())
+    } else {
+        Err(ErrorCode::SIZE)
+    }
+}
+```
+
+Similarly, avoid code that could divide by 0, avoid using `unwrap()` / `expect()`,
+and make sure to give the compiler enough information that it can guarantee `copy_from_slice()`
+is only being called on two slices of equal length.
+
+
+#### Non-generic-inner-functions
```

Comment:
Do we want to move "Non-generic-inner-functions" to be right after the heading on picking generics versus `dyn`? They seem closely related and perhaps they belong together.